executor/src/witgen/jit/processor.rs

```diff
@@ -0,0 +1,406 @@
+        )?;
+        let right_branch_code =
+            self.generate_code_for_branch(can_process, second_branch, complete, branch_depth + 1)?;
+        let code = if left_branch_code == right_branch_code {
+            common_code.into_iter().chain(left_branch_code).collect()
+        } else {
+            common_code
+                .into_iter()
+                .chain(std::iter::once(Effect::Branch(
+                    condition,
+                    left_branch_code,
+                    right_branch_code,
+                )))
+                .collect()
+        };
+
+        Ok(code)
+    }
+
+    fn process_until_no_progress<CanProcess: CanProcessCall<T> + Clone>(
+        &self,
+        can_process: CanProcess,
+        witgen: &mut WitgenInference<'a, T, FixedEval>,
+        complete: &mut HashSet<(u64, i32)>,
+    ) {
+        let mut progress = true;
+        while progress {
+            progress = false;
+
+            // TODO At this point, we should call a function on `witgen`
+            // to propagate known concrete values across the identities
+            // to other known (but not concrete) variables.
+
+            for (id, row_offset) in &self.identities {
+                if complete.contains(&(id.id(), *row_offset)) {
+                    continue;
+                }
+                let result = witgen.process_identity(can_process.clone(), id, *row_offset);
+                progress |= result.progress;
+                if result.complete {
+                    complete.insert((id.id(), *row_offset));
+                }
+            }
+        }
+    }
+
+    /// If any machine call could not be completed, that's bad because machine calls typically have side effects.
+    /// So, the underlying lookup / permutation / bus argument likely does not hold.
+    /// This function checks that all machine calls are complete, at least for a window of <block_size> rows.
+    /// It returns the list of incomplete calls, if any.
+    fn incomplete_machine_calls(&self, complete: &HashSet<(u64, i32)>) -> Vec<(&Identity<T>, i32)> {
+        self.identities
+            .iter()
+            .map(|(id, _)| id)
+            .filter(|id| is_machine_call(id))
+            .unique()
+            .flat_map(|&call| {
+                let rows = self.rows_for_identity(call);
+                let complete_rows = rows
+                    .iter()
+                    .filter(|&&row| complete.contains(&(call.id(), row)))
+                    .collect::<Vec<_>>();
+                // Because we process rows -1..block_size+1, it is fine to have two incomplete machine calls,
+                // as long as <block_size> consecutive rows are complete.
+                if complete_rows.len() >= self.block_size {
+                    let (min, max) = complete_rows.iter().minmax().into_option().unwrap();
+                    // TODO instead of checking for consequitive rows, we could also check
+                    // that they "fit" the next block.
```

Comment:
Are you saying that there could be gaps? Why? Eventually, we want the machine calls to be handled on all rows of the trace, right? The call being "complete" does not mean that it actually executed, it could be that the selector was 0, in which case it is also considered complete.

Reply:
But if the selector is zero, it will also be zero in the other, overlapping block. What I'm saying is that we can have
`B A B B B C B C C C D C D D D ...`
i.e. there is a gap of one that is filled by the previous / next block.

We should verify that with real machines, but I think the number of complete rows should be exactly block_size and row `i` should be complete if and only if row `i + block_size` is not complete.

Reply:
OK, sure ^^ Seems very esoteric though.